languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java

```diff
@@ -2247,80 +2266,76 @@ protected boolean ignorePotentiallyMisspelledWord(String word) throws IOExceptio
-          !isMisspelled(compound1) || ignorePotentiallyMisspelledWord(compound1) ||
-          !isMisspelled(compound1noS) || ignorePotentiallyMisspelledWord(compound1noS);
+
+      // Make sure that the individual parts of the compound have appropriate length.
+      // Short words can also be typos.
+      if (!isValidPartLength(part1, part2)) {
+        return false;
       }
-      boolean compound2ok =
-        (!isMisspelled(compound2) || ignorePotentiallyMisspelledWord(compound2)) && isNoun(compound2);
-      return compound1ok && compound2ok;
+
+      return processTwoPartCompounds(part1, part2);
+
+    } else if (parts.size() == 3) {
+      return processThreePartCompound(parts);
     } else {
       // more than three parts can be supported later
       return false;
     }
-    if (word.contains("-" + part2)) {
-      // don't accept e.g. "Implementierungs-pflicht"
-      return false;
-    }
+  }
+
+  private boolean isValidWordLength(String word) {
+    return word.length() <= MIN_WORD_LENGTH || word.length() >= MAX_WORD_LENGTH;
+  }
+
+  private boolean isValidPartLength(String part1, String part2) {
     // don't assume very short parts (like "Ei") are correct, these can easily be typos:
-    if ((hasInfixS || part1.endsWith("s")) && part1.length() >= 4 /* includes 's' */ && part2.length() >= 3 && startsWithLowercase(part2)) {
-      String part1noInfix = part1.substring(0, part1.length()-1);
-      String part2uc = uppercaseFirstChar(part2);
-      if ((compoundPatternWithHeit.matcher(part1).matches() || wordsNeedingInfixS.contains(part1noInfix)) &&
-          isNoun(part2uc)) {
-        if (compoundPatternWithAction.matcher(part1noInfix).matches() ||
-            compoundPatternWithFirst.matcher(part2uc).matches() ||
-            part1.endsWith("schwungs") || part1.endsWith("sprungs") || isMisspelled(part1noInfix) || isMisspelled(part2uc)) {
-          return false;
-        }
-        return true;
-      }
-    }
-    String part2uc = uppercaseFirstChar(part2);
-    if (!hasInfixS &&
-        part1.length() >= 3 && part2.length() >= 4 &&
-        !part2.contains("-") &&
-        startsWithLowercase(part2) &&
-        !part1.equals("Lass") &&  // e.g. "Lasstest" - couldn't find a more generic solution yet
-        (wordsWithoutInfixS.contains(part1) || (compoundPatternSpecialEnding.matcher(part1).matches() && isNoun(part2uc))) &&
-        !isMisspelled(part1) &&
-        isNoun(part2uc) // don't accept e.g. "Azubikommt"
-      ) {
-      System.out.println("compound: " + part1 + " " + part2 + " (" + word + ")");
-      return true;
-    }
-    return false;
+    return part1.length() >= 3 && part2.length() >= 4;
+  }
+
+  private String removeInfixS(String part1) {
+    return part1.replaceFirst("s?$", "");
+  }
+
+  private String removeHyphen(String part1) {
+    return part1.replaceFirst("-?$", "");
+  }
+
+  private String removeInfixSAndHyphen(String part1) {
+    return removeInfixS(removeHyphen(part1));
```

Comment:
🥜 Two replaces, when you could just be replacing `s?-?$` with `""` once in a single call. The performance cost is negligible (so from that point of view it's a nitpick), but do note that if there's a problem with `removeHyphen`, `removeInfixS` may not apply as expected.